storage: Handle timestamp collisions in timestampCache

bdarnell opened the pull request:
When two intervals in the timestamp cache have the same timestamp,
neither of them can be said to own that timestamp. We must adjust
intervals and clear transaction IDs whenever this collision occurs.

Failure to do so allowed one transaction to write after another
transaction had read at the same timestamp, leading to a violation of
serializability.

Fixes #9083

@cockroachdb/stability @nvanbenschoten

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9100)

<!-- Reviewable:end -->

bdarnell committed:
storage: timestampCache returns transaction IDs

timestampCache.GetMax{Read,Write} previously took a transaction ID to
act as that transaction. However, this wasn't always correct - when
acting as a transaction we look past that transaction's writes to see
what's underneath, but the trimming of inserted spans means that this
would often return the low water mark, ignoring spans that had been
present but were trimmed away.

Instead, the timestamp cache now returns the transaction associated with
the timestamp it is returning, so the caller can make its own decision
about whether to ignore the timestamp or not.

bdarnell committed:
storage: Handle timestamp collisions in timestampCache

When two intervals in the timestamp cache have the same timestamp,
neither of them can be said to own that timestamp. We must adjust
intervals and clear transaction IDs whenever this collision occurs.

Failure to do so allowed one transaction to write after another
transaction had read at the same timestamp, leading to a violation of
serializability.

Fixes #9083

nvanbenschoten commented:
Reviewed 4 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[storage/timestamp_cache.go, line 436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqcIdgazLFu7aJS5eB:-KQqcIdhYPgFkndML8Xu:b-5zdng5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L436)):*
> ```Go
> 		end = start.Next()
> 	}
> 	var ok bool
> ```

You didn't touch this, but we might want to improve the name of this flag.

---

*[storage/timestamp_cache.go, line 442 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqeg-Axe7KMBlInB3F:-KQqeg-Axe7KMBlInB3G:b-82wpwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L442)):*
> ```Go
> 		cache = tc.rCache
> 	}
> 	var txnID *uuid.UUID
> ```

`s/txnID/maxTxnID/` or something a bit more descriptive. Also move up next to `max := ...`

---

*[storage/timestamp_cache.go, line 451 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqcpO0DgJo1d4PW3GN:-KQqcpO0DgJo1d4PW3GO:bkrur8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L451)):*
> ```Go
> 		} else if max.Equal(ce.timestamp) && txnID != nil &&
> 			(ce.txnID == nil || !uuid.Equal(*txnID, *ce.txnID)) {
> 			txnID = nil
> ```

It would be nice to alias `nil *UUID`s to a variable in this file to make checks against this state more explicit and to make [calls to timestampCache.add with a nil txn](https://github.com/cockroachdb/cockroach/blob/ecf0c0997501e7d36ea7b09d8e8b08babfe65cb0/storage/timestamp_cache.go#L405) more clear:

```
var anyTxnID *uuid.UUID = nil // or some other name
```

---

*[storage/timestamp_cache.go, line 223 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqh1mSMmTDfFc859Oo:-KQqh1mTyt76lACeidbw:b-whbzc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L223)):*
> ```Go
> 			sCmp := r.Start.Compare(key.Start)
> 			eCmp := r.End.Compare(key.End)
> 			if cv.timestamp.Less(timestamp) {
> ```

It would be nice to have a `roachpb.Timestamp.Compare(roachpb.Timestamp) int` method and then compare against `< 0, > 0, == 0`.

---

*[storage/timestamp_cache.go, line 337 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqhz77L2uiXZxNW_Ka:-KQqhz77L2uiXZxNW_Kb:b-icdbkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L337)):*
> ```Go
> 				switch {
> 				case sCmp == 0 && eCmp == 0:
> 					// New and old are equal; replace old with new and avoid the
> ```

`s/; replace old with new and avoid the need to insert new//`

---

*[storage/timestamp_cache.go, line 346 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqhsn_84cTKZq5BFV_:-KQqhsn_84cTKZq5BFVa:b4it8m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L346)):*
> ```Go
> 					// Nil: ------------
> 					clearTxnIfDifferent(&cv.txnID, txnID)
> 					tcache.MoveToEnd(entry)
> ```

This shouldn't be needed, because the entry isn't getting a new timestamp, right?

---

*[storage/timestamp_cache.go, line 370 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqj9cF8vlpvwBoBRcM:-KQqj9cF8vlpvwBoBRcN:buys0ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L370)):*
> ```Go
> 					clearTxnIfDifferent(&cv.txnID, txnID)
> 					r.End = key.Start
> 					addRange(r)
> ```

Is:
```
addRange(r)
return
```
needed here? This should be the last overlap.

---

*[storage/timestamp_cache.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqjiIIglTb74j5MRIo:-KQqjiIIglTb74j5MRIp:bpvj34e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L383)):*
> ```Go
> 					clearTxnIfDifferent(&cv.txnID, txnID)
> 					newKey := tcache.MakeKey(r.Start, key.Start)
> 					newEntry := makeCacheEntry(newKey, cacheValue{timestamp: timestamp, txnID: txnID})
> ```

Can we define this `cacheValue` above `addRange := func(r interval.Range) {` and use in both places?

---

*[storage/timestamp_cache.go, line 409 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQql-xncEVol3Grn2n0:-KQql-xoqJrAwJpNEtjh:b-r6bi9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L409)):*
> ```Go
> 					// Old: ------------
> 					//
> 					// New:
> ```

Nit: I like how this diagram defines all three (New, Nil, and Old). Could we do that for all (maybe with Nil using a different symbol like `====`)?

---

*[storage/timestamp_cache.go, line 425 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqmmOY4tI006DrWeH2:-KQqmmOZP5ffGmcdi0Lq:bvm9w8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L425)):*
> ```Go
> 					// Old: ----
> 					key.End, r.Start = r.Start, key.End
> 					key := tcache.MakeKey(key.End, r.Start)
> ```

Can we rename this variable, it confused me for a little while because I didnt see the `:=`

---

*[storage/timestamp_cache.go, line 426 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqmCM1VBMCQ0ZDGm6T:-KQqmCM2lyf4OjEq66bY:ba133qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L426)):*
> ```Go
> 					key.End, r.Start = r.Start, key.End
> 					key := tcache.MakeKey(key.End, r.Start)
> 					newCV := cacheValue{timestamp: cv.timestamp, txnID: txnID}
> ```

Again, we might want to define this above? See comment on line 383. Here we could just reassign the value if we need to mutate it.

---

*[storage/timestamp_cache.go, line 431 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqlcG6okRPxQ7MsFQv:-KQqlcG6okRPxQ7MsFQw:b-3nrf1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L431)):*
> ```Go
> 					tcache.AddEntryAfter(newEntry, entry)
> 				case sCmp == 0:
> 					// Left-aligned partial overlap; truncate old start and
> ```

Here and in a few other states (`Right-aligned partial overlap`, `Left partial overlap`, `Right partial overlap`), if the TxnIDs are the same, we can return without splitting intervals, right (or only doing one split)? This is similar to what we do with `New and old are equal`. 

This means we might want `clearTxnIfDifferent` to return a flag signifying if neither UUID is nil and if both a equal.

---

*[storage/timestamp_cache.go, line 458 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqn86P3ZxmTGNmArRg:-KQqn86QCW0dTc5t2CwX:b-gwbs8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L458)):*
> ```Go
> 					tcache.AddEntryAfter(newEntry, entry)
> 					// We can add the new range now because it is guaranteed to
> 					// be any other overlaps; we ust do so because we've changed
> ```

`s/ust/must`

---

*[storage/timestamp_cache.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqnLMeEcYiBHqdMj-O:-KQqnLMfHDBr3VHMzlAq:b-qgnmio) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L460)):*
> ```Go
> 					// be any other overlaps; we ust do so because we've changed
> 					// our boundaries and continuing to iterate may hit the "no
> 					// overlap" panic.
> ```

Really? Won't this be guaranteed to be the last overlap?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
For my own edification, would you mind explaining why we need to nil out the span's owner rather than letting the transaction that got there first maintain ownership of it? What would be the harm in letting the "first" transaction finish as long as we prevented the other transactions from continuing with the same timestamp?

bdarnell committed:
storage: timestampCache PR feedback

bdarnell committed:
storage: Optimize timestampCache for equal timestamps and txns

When intervals have both the same timestamp and transaction ID
(which is fairly common since this is what happens whenever
a transaction touches the same rows more than once), we can avoid
the splitting required when timestamps are equal but transaction IDs
differ. Introduce (simple) special cases for this scenario.

bdarnell commented:
Review status: 3 of 4 files reviewed at latest revision, 14 unresolved discussions.

---

*[storage/timestamp_cache.go, line 436 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqcIdgazLFu7aJS5eB:-KQsmdghwxtYtW1UwIx2:bsd7i13) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L436)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> You didn't touch this, but we might want to improve the name of this flag.
</details>
`ok` is a fairly go idiomatic name for this kind of thing, but I've expanded the comments about it.

---

*[storage/timestamp_cache.go, line 442 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqeg-Axe7KMBlInB3F:-KQsmjValwV-zWPRWo8s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L442)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/txnID/maxTxnID/` or something a bit more descriptive. Also move up next to `max := ...`
</details>
Done.

---

*[storage/timestamp_cache.go, line 451 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqcpO0DgJo1d4PW3GN:-KQsm317ru2V_NowiQ_M:bi3dzoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L451)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> It would be nice to alias `nil *UUID`s to a variable in this file to make checks against this state more explicit and to make [calls to timestampCache.add with a nil txn](https://github.com/cockroachdb/cockroach/blob/ecf0c0997501e7d36ea7b09d8e8b08babfe65cb0/storage/timestamp_cache.go#L405) more clear:
> 
> ```
> var anyTxnID *uuid.UUID = nil // or some other name
> ```
</details>
I don't think that defining aliases for nil like that generally improve clarity. 

---

*[storage/timestamp_cache.go, line 223 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqh1mSMmTDfFc859Oo:-KQsgkuGoJ44_46VupZf:bf95kyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L223)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> It would be nice to have a `roachpb.Timestamp.Compare(roachpb.Timestamp) int` method and then compare against `< 0, > 0, == 0`.
</details>
Maybe, but I'm not sure if we'd use it anywhere else and for a single use this is fine.

---

*[storage/timestamp_cache.go, line 337 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqhz77L2uiXZxNW_Ka:-KQsh5a9PASN0HHb-yiZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L337)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/; replace old with new and avoid the need to insert new//`
</details>
Done.

---

*[storage/timestamp_cache.go, line 346 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqhsn_84cTKZq5BFV_:-KQsh8mHA1z4pNnEsU7c:b-n2k3k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L346)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This shouldn't be needed, because the entry isn't getting a new timestamp, right?
</details>
Right. Removed.

---

*[storage/timestamp_cache.go, line 370 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqj9cF8vlpvwBoBRcM:-KQshH8VU4kmDagj9TWq:bd88qew) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L370)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Is:
> ```
> addRange(r)
> return
> ```
> needed here? This should be the last overlap.
</details>
I don't think it is; removed.

---

*[storage/timestamp_cache.go, line 383 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqjiIIglTb74j5MRIo:-KQshOWUiaVyazb6tXLR:b-j1mye9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L383)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Can we define this `cacheValue` above `addRange := func(r interval.Range) {` and use in both places?
</details>
I think we need the late binding of `txnID` in cases where there are multiple overlaps. 

---

*[storage/timestamp_cache.go, line 409 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQql-xncEVol3Grn2n0:-KQshcz_BpodiuvQgU0D:bem0lzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L409)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Nit: I like how this diagram defines all three (New, Nil, and Old). Could we do that for all (maybe with Nil using a different symbol like `====`)?
</details>
OK, filled in all the example outputs even when they're the same as the inputs or empty. 

---

*[storage/timestamp_cache.go, line 425 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqmmOY4tI006DrWeH2:-KQshmEW-mPnWAL9z2pR:b3ymwd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L425)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Can we rename this variable, it confused me for a little while because I didnt see the `:=`
</details>
Yeah. I copied this pattern from some existing code but got tripped up by the shadowing at one point too. I'll rename them all (and I'm surprised go vet didn't complain)

---

*[storage/timestamp_cache.go, line 431 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqlcG6okRPxQ7MsFQv:-KQsvfD5YilyGV4qkmo3:b-16p2pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L431)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Here and in a few other states (`Right-aligned partial overlap`, `Left partial overlap`, `Right partial overlap`), if the TxnIDs are the same, we can return without splitting intervals, right (or only doing one split)? This is similar to what we do with `New and old are equal`. 
> 
> This means we might want `clearTxnIfDifferent` to return a flag signifying if neither UUID is nil and if both a equal.
</details>
Good point. It's just an optimization but it does seem like a common one since a transaction that touches the same rows multiple times will hit this case. Instead of returning a value from clearTxnIfDifferent I've made a whole separate switch statement with it's own ascii-diagrams for the case when both timestamps and transaction IDs are the same. 

---

*[storage/timestamp_cache.go, line 458 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqn86P3ZxmTGNmArRg:-KQsjpUjbaEgQSmlMOfB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L458)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/ust/must`
</details>
Done.

---

*[storage/timestamp_cache.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KQqnLMeEcYiBHqdMj-O:-KQsl7n66EvsrI7EGNV9:bvgb38o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L460)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Really? Won't this be guaranteed to be the last overlap?
</details>
Yeah, I think the panics I was seeing before were from other bugs. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Would you mind explaining why we need to nil out the span's owner rather than letting the transaction that got there first maintain ownership of it? What would be the harm in letting the "first" transaction finish as long as we prevented the other transactions from continuing with the same timestamp? I assume there will be cases in which the "first" transaction is already done by the time the "second" one sees the cache entry anyway.

---

Review status: 3 of 4 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-:-KQzjJnVnzXY-YB98hJW:bw1eps9)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
In this case the two transactions are read-only at the time of the conflict (they go on to write later). We can't require that all read-only transactions have unique timestamps (think of time-travel queries), so we have to allow the reads to proceed and deal with the overlap if and only if they later issue a conflicting write.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I only looked at some of the same-txn optimizations in detail.

---

Reviewed 4 of 4 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[storage/timestamp_cache.go, line 416 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KR-0aloETgLjELTACQc:-KR-0aloETgLjELTACQd:b-sksc45) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d93b1e0e23cfdc4c78877129afd038748c895ec/storage/timestamp_cache.go#L416)):*
> ```Go
> // transactions in the cache, the low water timestamp is returned for
> // the read timestamps. Also returns an "ok" bool, indicating whether
> // an explicit match of the interval was found in the cache.
> ```

From the comment, I'm not clear on what the last part means. `ok` is equivalent to `uuid != nil` or when the timestamp returned didn't originate from the low water mark?

---

*[storage/timestamp_cache.go, line 339 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KR-4Wy20WTimeDL8h5U:-KR-4Wy20WTimeDL8h5V:bby7n6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6873b97c43eb9683780fff14b3938f82b61b333/storage/timestamp_cache.go#L339)):*
Should this read

> Segment is unowned if multiple transactions were involved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-:-KR-5byrkjVqszagI5Fg:b-1tuht9)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/timestamp_cache.go, line 354 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KR-fXOqFFvTMjtMsfFq:-KR-fXOqFFvTMjtMsfFr:bazwaki) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca72c6b78506c08cc4fab8a8ba73af7eeb3ca5a/storage/timestamp_cache.go#L354)):*
> ```Go
> 					// Old:
> 					tcache.DelEntry(entry)
> 				case eCmp >= 0:
> ```

For this case and the next, cant we just delete old and extend new's bounds to avoid multiple entries?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-:-KR-fs75rPcPdcmiUwFb:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/timestamp_cache.go, line 354 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KR-fXOqFFvTMjtMsfFq:-KR1EW8FNDRrXZUcT6TW:b-7hemtq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca72c6b78506c08cc4fab8a8ba73af7eeb3ca5a/storage/timestamp_cache.go#L354)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> For this case and the next, cant we just delete old and extend new's bounds to avoid multiple entries?
</details>
That would be the first time we expand the bounds of an interval rather than shrinking it. Is it safe to do so? Why expand new instead of old? If it's safe to expand an interval, then we could replace this whole switch (the outcome of every case is a single span covering the union of old and new). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/timestamp_cache.go, line 354 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100#-KR-fXOqFFvTMjtMsfFq:-KR1UPgOZ-Eht2UDm2pJ:b-l2e2me) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca72c6b78506c08cc4fab8a8ba73af7eeb3ca5a/storage/timestamp_cache.go#L354)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> That would be the first time we expand the bounds of an interval rather than shrinking it. Is it safe to do so? Why expand new instead of old? If it's safe to expand an interval, then we could replace this whole switch (the outcome of every case is a single span covering the union of old and new). 
</details>
Yeah you're right, that would be a first. I dont think it's worth making this logic any more complex.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9100)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
storage: Adapt to removal of uuid.Equal
